src/librustc/mir/interpret/allocation.rs

```diff
@@ -101,18 +117,62 @@ impl<Tag> Allocation<Tag> {
     }
 
     pub fn undef(size: Size, align: Align) -> Self {
+        // Ensure initialization later gives the correct length.
         assert_eq!(size.bytes() as usize as u64, size.bytes());
         Allocation {
-            bytes: vec![0; size.bytes() as usize],
+            alloc: None,
             relocations: Relocations::new(),
-            undef_mask: UndefMask::new(size, false),
+            size,
             align,
             mutability: Mutability::Mutable,
             extra: (),
         }
     }
 }
 
+/// Raw accessors. Provide compatibility through lazy allocation of undef.
+impl<Tag, Extra> Allocation<Tag, Extra> {
+    pub fn len(&self) -> usize {
+        self.alloc.as_ref()
```

Comment:
you can just make this function body `self.size.bytes() as usize`, the `alloc.bytes` debug check is just a sanity check after all